src/java_tools/buildjar/java/com/google/devtools/build/buildjar/javac/BlazeJavaCompiler.java

```diff
@@ -81,6 +91,35 @@ public JavaCompiler make(Context c) {
         });
   }
 
+  @Override
+  public void initProcessAnnotations(
```

Comment:
Would it work to handle the `UnsupportedClassVersionError` at a higher level, instead of overriding `initProcessAnnotations` and needing another `--add-exports=` flag? This class is already making extensive use of internal APIs, but it's nice to avoid adding more uses if we can help it.

I made a quick attempt at handling it in `BlazeJavacMain` and it passes the attached test:

```diff
diff --git a/src/java_tools/buildjar/java/com/google/devtools/build/buildjar/javac/BlazeJavacMain.java b/src/java_tools/buildjar/java/com/google/devtools/build/buildjar/javac/BlazeJavacMain.java
index 0ddc57a2ce..0a460ab0b4 100644
--- a/src/java_tools/buildjar/java/com/google/devtools/build/buildjar/javac/BlazeJavacMain.java
+++ b/src/java_tools/buildjar/java/com/google/devtools/build/buildjar/javac/BlazeJavacMain.java
@@ -54,6 +54,8 @@ import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 import javax.annotation.Nullable;
 import javax.tools.Diagnostic;
 import javax.tools.DiagnosticListener;
@@ -70,6 +72,11 @@ import javax.tools.StandardLocation;
  */
 public class BlazeJavacMain {

+  private static final Pattern UNSUPPORTED_CLASS_VERSION_ERROR =
+      Pattern.compile(
+          "^(?<class>[^ ]*) has been compiled by a more recent version of the Java Runtime "
+              + "\\(class file version (?<version>[4-9][0-9])\\.");
+
   /**
    * Sets up a BlazeJavaCompiler with the given plugins within the given context.
    *
@@ -139,8 +146,24 @@ public class BlazeJavacMain {
         throw e.getCause();
       }
     } catch (Exception t) {
-      if (t.getCause() instanceof CancelRequestException) {
-        return BlazeJavacResult.cancelled(t.getCause().getMessage());
+      Throwable cause = t.getCause();
+      if (cause instanceof CancelRequestException) {
+        return BlazeJavacResult.cancelled(cause.getMessage());
+      }
+      Matcher matcher;
+      if (cause instanceof UnsupportedClassVersionError
+          && (matcher = UNSUPPORTED_CLASS_VERSION_ERROR.matcher(cause.getMessage())).find()) {
+        // Java 8 corresponds to class file major version 52.
+        int processorVersion = Integer.parseUnsignedInt(matcher.group("version")) - 44;
+        errWriter.println(
+            String.format(
+                "The Java %d runtime used to run javac is not recent enough to run the processor %s, "
+                    + "which has been compiled targeting Java %d. Either register a Java toolchain "
+                    + "with a newer java_runtime or, if this processor has been built with Bazel, "
+                    + "specify a lower --tool_java_language_version.",
+                Runtime.version().feature(),
+                matcher.group("class").replace('/', '.'),
+                processorVersion));
       }
       t.printStackTrace(errWriter);
       status = Status.CRASH;
```

Reply:
I was slightly worried about this message becoming a false positive in case javac throws `UnsupportedClassVersionError` for classes that aren't annotation processors. But I guess that's pretty unlikely as it would usually emit an error directly.

Reply:
Thanks! We can revisit if that ends up being an issue, but fingers crossed its much less likely to see `UnsupportedClassVersionError`s that aren't related to annotation processors.